File: dollarx/src/test/java/info/dollarx/BasicPathCreationTest.scala
package info.dollarx

import org.junit.Test
import org.junit.Assert.assertThat
import org.hamcrest.CoreMatchers.equalTo
import Path._
import ElementProperties._

class BasicPathCreationTest {

   @Test def divBeforeSpan {
    val el = div before span
    val xpath: String = el.getXPath.get
    assertThat(xpath, equalTo("span/preceding::div"))
    assertThat(el.toString(), equalTo("div, before span"))
  }

  @Test def divAfterSpan {
    val el = div that hasText("John") after span
    val xpath: String = el.getXPath.get
    assertThat(xpath, equalTo("span/following::div[translate(text(), 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz') = 'john']"))
    assertThat(el.toString, equalTo("""div, that has the text "John", after span"""))
  }

  @Test def divBeforeSiblingSpan {
    val el = div beforeSibling span
    val xpath: String = el.getXPath.get
    assertThat(xpath, equalTo("span/preceding-sibling::div"))
    assertThat(el.toString, equalTo("div, before the sibling span"))
  }

  @Test def complexBeforeAndAfter {
    val el: Path = (div that isBefore(span)).before(span that isAfter(div that isBefore(span)))
    val xpath: String = el.getXPath.get
    assertThat(xpath, equalTo("span[preceding::div[following::span]]/preceding::div[following::span]"))
    assertThat(el.toString, equalTo("div, that is before: span, before (span, that is after: (div, that is before: span))"))
  }

  @Test def withClassAndIndex {
    val el: Path = (div that hasClass("foo"))(5)
    val xpath: String = el.getXPath.get
    assertThat(xpath, equalTo("(//div[contains(concat(' ', normalize-space(@class), ' '), ' foo ')])[6]"))
    assertThat(el.toString, equalTo("""occurrence number 6 of (div, that has class "foo")"""))
  }

  @Test def withClassAndInside {
    val dialog: Path = div withClass "ui-dialog"
    val el: Path = span that(has cssClass "foo", is inside dialog)
    val xpath: String = el.getXPath.get
    assertThat(xpath, equalTo("span[contains(concat(' ', normalize-space(@class), ' '), ' foo ')][ancestor::div[contains(concat(' ', normalize-space(@class), ' '), ' ui-dialog ')]]"))
    assertThat(el.toString, equalTo("""span, that has class "foo", and has ancestor: (div, that has class "ui-dialog")"""))
  }

  @Test def withClassAndInsidedialogWithDescription {
    val dialog: Path = (div withClass "ui-dialog") describedBy "sumbission form"
    val el: Path =  span that(has cssClass "foo" , is inside dialog)
    val xpath: String = el.getXPath.get
    assertThat(xpath, equalTo("span[contains(concat(' ', normalize-space(@class), ' '), ' foo ')][ancestor::div[contains(concat(' ', normalize-space(@class), ' '), ' ui-dialog ')]]"))
    assertThat(el.toString, equalTo("""span, that has class "foo", and has ancestor: (sumbission form)"""))
  }

  @Test def divOrSpan {
    val el: Path = div or span
    val xpath: String = el.getXPath.get
    assertThat(xpath, equalTo("*[self::div | self::span]"))
    assertThat(el.toString, equalTo("div or span"))
  }

  @Test def divWithClassOrSpan {
    val el: Path = (div withClass "foo") or span
    val xpath: String = el.getXPath.get
    assertThat(xpath, equalTo("*[self::div[contains(concat(' ', normalize-space(@class), ' '), ' foo ')] | self::span]"))
    assertThat(el.toString,  equalTo("""(div, that has class "foo") or span"""))
  }
}

File: jdollarx/src/test/java/com/github/loyada/jdollarx/BasicPathCreationTest.java
package com.github.loyada.jdollarx;

import org.junit.Test;

import static com.github.loyada.jdollarx.BasicPath.div;
import static com.github.loyada.jdollarx.BasicPath.span;
import static org.junit.Assert.assertThat;
import static org.hamcrest.CoreMatchers.*;
import static com.github.loyada.jdollarx.ElementProperties.*;

public class BasicPathCreationTest {

    @Test
    public void divBeforeSpan(){
        Path el = div.before(span);
        String xpath = el.getXPath().get();
        assertThat(xpath, is(equalTo("span/preceding::div")));
        assertThat(el.toString(), is(equalTo("div, before span")));
    }

    @Test
    public void divAfterSpan(){
        Path el = div.after(span);
        String xpath = el.getXPath().get();
        assertThat(xpath, is(equalTo("span/following::div")));
        assertThat(el.toString(), is(equalTo("div, after span")));
    }

    @Test
    public void divBeforeSiblingSpan(){
        Path el = div.beforeSibling(span);
        String xpath = el.getXPath().get();
        assertThat(xpath, is(equalTo("span/preceding-sibling::div")));
        assertThat(el.toString(), is(equalTo("div, before the sibling span")));

    }

    @Test
      public void complexBeforeAndAfter(){
        // this is a really complicated way of saying div.before(span), for the sake of testing
        Path el = div.that(isBefore(span)).before(span.that(isAfter(div.that(isBefore(span)))));
        String xpath = el.getXPath().get();
        assertThat(xpath, is(equalTo("span[preceding::div[following::span]]/preceding::div[following::span]")));
        assertThat(el.toString(), is(equalTo("div, that is before: span, before (span, that is after: (div, that is before: span))")));
    }

    @Test
    public void withClassAndIndex(){
        Path el = div.that(hasClass("foo")).withGlobalIndex(5);
        String xpath = el.getXPath().get();
        assertThat(xpath, is(equalTo("(//div[contains(concat(' ', normalize-space(@class), ' '), ' foo ')])[6]")));
        assertThat(el.toString(), is(equalTo("occurrence number 6 of (div, that has class foo)")));
    }

    @Test
    public void withClassAndInside(){
        Path dialog = div.withClass("ui-dialog");
        Path el = span.that(hasClass("foo"), isInside(dialog));
        String xpath = el.getXPath().get();
        assertThat(xpath, is(equalTo("span[contains(concat(' ', normalize-space(@class), ' '), ' foo ')][ancestor::div[contains(concat(' ', normalize-space(@class), ' '), ' ui-dialog ')]]")));
        assertThat(el.toString(), is(equalTo("span, that has class foo, and has ancestor: (div, that has class ui-dialog)")));
    }

    @Test
    public void withClassAndInsidedialogWithDescription(){
        Path dialog = div.withClass("ui-dialog").describedBy("sumbission form");
        Path el = span.that(hasClass("foo"), isInside(dialog));
        String xpath = el.getXPath().get();
        assertThat(xpath, is(equalTo("span[contains(concat(' ', normalize-space(@class), ' '), ' foo ')][ancestor::div[contains(concat(' ', normalize-space(@class), ' '), ' ui-dialog ')]]")));
        assertThat(el.toString(), is(equalTo("span, that has class foo, and has ancestor: (sumbission form)")));
    }

    @Test
    public void divOrSpan(){
        Path el = div.or(span);
        String xpath = el.getXPath().get();
        assertThat(xpath, is(equalTo("*[(self::div) | (self::span)]")));
        assertThat(el.toString(), is(equalTo("div or span")));
    }

    @Test
    public void divWithClassOrSpan(){
        Path el = div.withClass("foo").or(span);
        String xpath = el.getXPath().get();
        assertThat(xpath, is(equalTo("*[(self::div[contains(concat(' ', normalize-space(@class), ' '), ' foo ')]) | (self::span)]")));
        assertThat(el.toString(), is(equalTo("(div, that has class foo) or span")));
    }

}
